pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+				}
+
+				toRepo, err := o.Repository(ctx, toContext, creds, dst.t, dst.ref)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to connect to %s: %v", dst.ref, err)})
+					continue
+				}
+
+				canonicalTo := toRepo.Named()
+				toManifests, err := toRepo.Manifests(ctx)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to access destination image %s manifests: %v", src.ref, err)})
+					continue
+				}
+
+				// if the destination tag already has this manifest, do nothing
+				var mustCopyLayers bool
+				if o.Force {
+					mustCopyLayers = true
+				} else {
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
+						default:
+							digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("the manifest type %T is not supported", srcManifest)})
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							// tagging within the same registry is a no-op
+							if src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String() {
+								continue
+							}
+
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+
+							// if we aren't forcing upload, skip the blob copy
+							if !o.Force {
+								_, err := toBlobs.Stat(ctx, blob.Digest)
+								if err == nil {
+									// blob exists, skip
+									glog.V(5).Infof("Server reports blob exists %#v", blob)
+									continue
+								}
+								if err != distribution.ErrBlobUnknown {
+									glog.V(5).Infof("Server was unable to check whether blob exists %s: %v", blob.Digest, err)
+								}
+							}
+
+							var options []distribution.BlobCreateOption
+							if !o.SkipMount {
+								options = append(options, client.WithMountFrom(blobSource), WithDescriptor(blob))
+							}
+							w, err := toBlobs.Create(ctx, options...)
+							// no-op
+							if err == ErrAlreadyExists {
+								glog.V(5).Infof("Blob already exists %#v", blob)
+								continue
+							}
+							// mount successful
+							if ebm, ok := err.(distribution.ErrBlobMounted); ok {
+								glog.V(5).Infof("Blob mounted %#v", blob)
+								if ebm.From.Digest() != blob.Digest {
```

Comment:
Not sure if it's worth it. Docker daemon doesn't check this.

Reply:
We are checking it because in a mirroring case we may have incorrect data, thus an admin may need to run --force.